tp030/README.md

```diff
@@ -0,0 +1,88 @@
+
+## Obtaining and Processing Grants
+When Bob receives the message from Alice-ChatApp, either the message contains the entire grant in the `delegatesGrant` field OR it contains a CID of the grant `DelegatesGrant` with. In the former case, if Bob has never seen the grant before, he will process and store it in his DWN. In the latter case, he must already have the grant in his DWN.
+
+## Revocation
+### Phase 1: `DelegatesCheck`
+When Bob authorizes a message with a `DelegatesGrant`, he will send a `DelegatesCheck` message to Alice's to check if the grant is still active. If the grant is active, Alice will say so. If the grant has been revoked, Alice will send the `DelegatesRevoke` message in the response along with a list of message CIDs to retain. To reduce network calls, Bob will cache the result of `DelegatesCheck`.
+
+This first phase only requires `DelegatesGrant`, `DelegatesRevoke`, and `DelegatesCheck` to exist.
+
+### Phase 2: `DelegatesRegister`
+We build on Phase 1 to further reduce network calls. The first time Bob sees a given `DelegatesGrant`, he will send a `DelegatesRegister` to Alice, adding Bob to a list of entities who want to know when a grant is revoked. If the grant has already been revoked when Bob sends the `DelegatesRegister`, he will receive the `DelegatesRevoke` and retention list in the response. After Bob sends the `DelegatesRegister`, he will not need to ping Alice about revocation again. When Alice wants to revoke Alice-ChatApp's grant, she will get the list of DIDs who sent a `DelegatesRegister` for that grant and send the `DelegatesRevoke` to them. If Bob is not able to receive the `DelegatesRevoke`, he will likely know in advance, leading him to use the Phase 1 approach.
+
+## Message structure
+```ts
+type DelegatesGrantDescriptor = {
+  messageTimestamp: string;
+  delegatedTo: string;
+  delegatedBy: string;
+}
+type DelegatesGrantReply = {
+  status: Status;
+};
+
+type DelegatesRevokeDescriptor = {
+  messageTimestamp: string;
+  delegatesGrantId: string;
+  retentionListId: string; // CID of retention list sent in data payload
+};
+type DelegatesRevokeReply = {
+  status: Status;
+};
+
+type DelegatesCheckDescriptor = {
+  messageTimestamp: string;
+  delegatesGrantId: string;
+};
+type DelegatesCheckResponse = {
+  status: Status;
+  revoke?: DelegatesRevokeMessage;
+  retentionlist?: string[];
+};
+
+type DelegatesRegisterDescriptor = {
+  messageTimestamp: string;
+  delegatesGrantId: string;
+}
+type DelegatesRegisterResponse = {
+  status: Status;
+  revoke?: DelegatesRevokeMessage;
+  retentionlist?: string[];
+};
+```
+
+## Why isn't this part of the `Permissions` interface?
+The two interfaces are conceptually distinct. `Permissions` is introverted; it controls whether an external entity can affect change on my DWN. `Delegates` is extroverted; it determines whether an external entity can affect change on other parties' DWNs in my name.
```

Comment:
Delegate grants are also introverted because it will also be used to write to my DWN. For instance, a chat app DID `alice-chat` starting a chat thread and writing messages into Alice's own DWN.